c_src/erasure.c

```diff
@@ -265,22 +381,214 @@ decode(ErlNifEnv * env, int argc, const ERL_NIF_TERM argv[])
+            goto cleanup;
+        }
+        lasttotalsize=totalsize;
+
+        if (blocksize == 0) {
+            blocksize = totalsize / k;
+            padding = 0;
+
+            remainder = totalsize % k;
+            if (remainder != 0) {
+                // payload is not cleanly divible by K, we need to pad
+                padding = (k - (totalsize % k));
+                blocksize = (totalsize + padding) / k;
+            }
+            while (blocksize % sizeof(long) != 0 || blocksize % w != 0) {
+                blocksize++;
+                padding++;
+            }
+            // block spacing has to be a multiple of 16
+            blockspacing = blocksize + (16 - (blocksize % 16));
+            // Note more random padding
+            shards = calloc(k+m+1, blockspacing);
+        }
+
+        ErlNifBinary input;
+        if (!enif_is_binary(env, tuple[2]) || !enif_inspect_binary(env, tuple[2], &input) || input.size != blocksize)
+        {
+            result = enif_make_tuple2(env, enif_make_atom(env, "error"), enif_make_atom(env, "invalid_shard"));
+            goto cleanup;
+        }
+
+        if (id < k) {
+            data_ptrs[id] = shards+(blockspacing*id);
+        } else {
+            coding_ptrs[id-k] = shards+(blockspacing*id);
+        }
+
+        memset(shards+(blockspacing*id), 0, blockspacing);
+        memcpy(shards+(blockspacing*id), input.data, blocksize);
+    }
+
+    erasures = calloc(k+m, sizeof(int));
+    int j = 0;
+
+    int bytes_per_shard = totalsize / k;
+    int extra_bytes = totalsize % k;
+
+    // calculate the missing shards and track them in the erasures array
+    for (int i = 0; i < k+m; i++) {
+        if ((i < k && data_ptrs[i] == NULL) || (i >= k && coding_ptrs[i-k] == NULL)) {
+            // set up any missing data or coding pointers
+            if (i < k && data_ptrs[i] == NULL) {
+                data_ptrs[i] = shards+(blockspacing*i);
+            } else if (i >= k && coding_ptrs[i-k] == NULL) {
+                coding_ptrs[i-k] = shards+(blockspacing*i);
+            }
+            erasures[j] = i;
+            j++;
+        }
+    }
+    erasures[j] = -1;
+
+    matrix = cauchy_good_general_coding_matrix(k, m, w);
+    if (matrix == NULL) {
+        result = enif_make_badarg(env);
+        goto cleanup;
+    }
+    bitmatrix = jerasure_matrix_to_bitmatrix(k, m, w, matrix);
+    if (bitmatrix == NULL) {
+        result = enif_make_badarg(env);
+        goto cleanup;
+    }
+    /* This works, but is slower */
+    /*int res = jerasure_bitmatrix_decode(k, m, w, bitmatrix, 0, erasures, data_ptrs, coding_ptrs, blocksize, blocksize/w);*/
+    /* This works but ONLY if you don't use smart mode (the last argument set to 1 uses smart mode)
+     * smart mode seems to mis-allocate memory for the smart schedule and that causes segfaults */
+    int res = jerasure_schedule_decode_lazy(k, m, w, bitmatrix, erasures, data_ptrs, coding_ptrs, blocksize, blocksize/w, 0);
```

Comment:
Isn't `jerasure_bitmatrix_decode` faster than `decode_lazy` when not using the smart mode?

Reply:
Let me test that

Reply:
Right, so, `jerasure_bitmatrix_decode` is indeed _slightly_ faster than the dumb mode `decode_lazy`.

Quick test:

```Decoded with jerasure_bitmatrix_decode - Bytes XOR'd: 6752.
jerasure_schedule_decode_lazy - Bytes XOR'd: 5696.
jerasure_schedule_decode_lazy_dumb - Bytes XOR'd: 6888.
```